Sliding freq in Audio RX

strijar opened the pull request:
I have implemented a sliding frequency in the audio receiver.
Demo - https://youtu.be/dqFCD6OgVCI?t=74

gullradriel commented:
My Russian is rusty, can you please explain what it is doing?

gullradriel commented:
From the video I understand that it is tuning on what we see without moving center freq.
Looks usefull, I will have a deeper checking next Monday

strijar commented:
Yes that's right. You can change the frequency inside the waterfall without changing it. Especially useful when listening to the telegraph

Brumi-2021 commented:
Congratulations, great job ! 
Good LUT colours! 
Looks very attractive  the colours and the sliding, willing to test !
I guess that when the want to tune a  freq out of LCD area , the sliding can be deactivated, or just shifted all contents  . Let's  check . 
Thanks!

strijar commented:
Yes, when you try to go beyond the spectrum, it shifts a little

Brumi-2021 commented:
Hi @strijar , can you please share your beta binary for testing (including both PR's) , through here zipped binary or through discord test-drive  direct binary ? 
73 , thanks

strijar commented:
[portapack-h1_h2-mayhem.zip](https://github.com/eried/portapack-mayhem/files/9995141/portapack-h1_h2-mayhem.zip)

Brumi-2021 commented:
Hello thanks @strijar for sharing it . 
Now I better understand  the LUT  colour change .
![image](https://user-images.githubusercontent.com/86470699/201485342-c5aced1e-60b9-48d1-ad14-ea118dc38b3e.png)

It is a matter of taste,  but I also agree that the new one seems more attractive and clear.
(I have seen , that it will be also be automatically applied to the Capture App waterfall.)
If we are not increasing too much the flash size (<1kB) , selectable LUT would be the best .
If not ,we will need to vote and decide , but personally I like more yours.


Regarding the top sliding tuning bar  issue , 
![image](https://user-images.githubusercontent.com/86470699/201485436-c3fac435-d040-4f7f-83cf-034933fa0253.png)

I also found it quite interesting,. I understand that it is only applied to the AM and NFB modes, not ?  (not to to the WFM , nor to the SPEC, not ?) .   Again it is a matter of UI and user experience , but I think that some good SDR programs it has it , and looks like to me , good feature , but in my quick test I found two issues, 

i-) two times after quick  tuning freq. change slide , steps 9khz of 25khz I managed to hand up the system two times  . (not sure , but maybe that hang up or instability was originally there ????) 
I could not find the steps to reproduce it .  ( I also need to come back to test in prev. fw and with other platform …- now just tested under  H2 WM8731 , big QFP.)  (Maybe not related to your PR ???) 

ii-) In current  FW , tuning a  AM DSB commercial broadcasting  with carrier , once good center  tuned,  the audio is ok  when changing AM to any SSB (LSB-USB) and viceversa.  In your FW ,in each demodulation change (AM ->LSB--USB)  , you need to retune freq one step up and later one step down, to recover again  a clean AM or SSB good audio . (it is a minor issue, because the user always can retune again a clean audio in any side lateral band . 
But it is strange and not logical compared to previous firmware  (this can be reproduce it 100% of times) 

I will try to continue testing , but more opinions and feedback will be  welcome !!!, 

But @strijar thanks for your contribution !!!
Cheers,

strijar commented:
The color palette for the waterfall is only 768 bytes, so you can create several to choose from. I made this palette for UHSDR firmware and there was originally a choice of 5 pieces (;

For now, it only works for narrowband signals. I tried for broadband, but it doesn't make sense with such a narrow waterfall. I've also tried extending the waterfall to 100-200kHz but there are performance issues so far. Need to work on optimization.

Yes, I do not touch the frequency shift when switching modulation. Probably correct to set it to 0

Brumi-2021 commented:
Hello @strijar (and all),  

I checked a little more your latest binary fw ,  with your two PR's ,
As you said ,  It is fine only apply the sliding freq. , just  to the narrow band RX (AM and NFMB), I agree with you. (no meaning  in WFM or SPEC) 
 
In my H2 - big QFP platform ,today  I also managed to randomly  hang up once again Audio RX App unders quick test (< 3 min)   , and  freeze the screen and not possible to recover by any button again. This happened  after several minutes tuning up and down in AM , using  MF up conversion .

In my  H1 platform , using exactly same configuration set up ,  all seems working fine .No hang up , all ok (under quick test) .

But as we have the other already pointed above  issue   (ii) :  in your FW ,at each demodulation change (AM ->LSB--USB) , you need to retune freq one step up and later one step down, to recover again a clean AM DSB or SSB good audio .) ,

Just as my personal opinion  :  
 I suggest to merge your first LUT PR , straight away . (I like it  very much !!!!)  

 But  regarding your second PR (sliding freq. Audio  RX)  ,it is also an attractive feature , but  in my personal opinion , is still not ready to be merged yet  ,  (current fw has a very  clean modulation type change AM-USB- LSB , without needing to make a retune 1 step up / 1 step down ) , in that sense , current fw is more friendly user  and logical . 

This is just my personal opinion,  but more  feedback testing and opinions will be welcome.
(Other option might be ,  merge your both PR's , but continue improving it  and optimizing the last one ) .

Cheers,  and congratulations for your great contributions !

strijar commented:
I agree, it's too early to merge. I'm currently working on an optimized version of DDC (based on CORDIC) and it might work out better. There is hope that it will work on waterfalls with a width of 100-200KHz

73!

strijar commented:
I have implemented DDC on CORDIC. This gave a significant increase in speed. DDC is now running at 96kHz, which allows the frequency to slide within 48kHz. The width of the waterfall has also been increased to 192KHz. And it is possible to reduce the width of the waterfall (make zoom)

[portapack-h1_h2-mayhem.zip](https://github.com/eried/portapack-mayhem/files/10031507/portapack-h1_h2-mayhem.zip)

Brumi-2021 commented:
Hi thanks ! 👍
Later I will check it . I am no familiar yet with this DDC CORDIC down convertion algorithms.  
So far i am finalising a wider FIR decim and channel filters to be  added  in current Receiver Audio App, wider AM DSB 9Khz bandwidth (current is 6k ) to support also better audio quality in commercial AM broadcasting. 
I hope that compatible with your DDC CORDIC changes . 
Cheers

strijar commented:
Ok. I changed the frequency reduction scheme a little. Now the decimators are in this order /4  /8 /8 /1 (was /8 /8 /4 /1).

gullradriel commented:
I find it nice and useful, I encourage you to bullet proof test it and commit all you need to here, so it can finally be reviewed and merged.
Tell us when you are feeling ready :-)

strijar commented:
@gullradriel From my side, everything is ready, I did not find any errors. Need a side view (;

Brumi-2021 commented:
Hi , I have been checking your latest DDC Corvic binary .
I respect a lot your job . I believe that behind that binary , sure that  there are a lot  of hours and debugging !!!! 

Looks promising algoritmus engine !!! I did not know about it .  It has some very good issues, but still some major  issues.
I am not specialist about it , I guess that  all this down conversion with DDC Corvic is used for both issues GUI  and FFT waterfall, and  also for the audio demodulation ?  I read in some internet papers that it is extremely fast and good for the CPU's but also produces some more errors , but somehow  a good compromise of accuracy versus speed

Other question, when you changed the decimation order , did you recheck if the pre-aliasing FIR filters decim1,2, ... are still ok ? Some filters are running at different FS than original design, not ?  (I think did not check , if we have the same good sensitivity and image rejection as before) . 

See my quick feedback evaluation.
**(+) good points ,** 
* streemly fast !
* very big visible freq.  band , much broader  than current !  (this  could be very good , but on the other side , in AM the central green bar , now  is almost not visible ... It is normal , +-3k compared to +-50k spectrum is less visible than when we were watching +-20k ...)   

**(-) points to be solved or improve ,** 
* I still managed to hang up several times in AM , WFM , several times  (that need a hard reset) (not stable  enough ) .
* the majority of times the cursor was showing correctly  a tuned broadcasting carrier matching with the central cursor , but after moving up and down , and inputing the same freq. by the keyboard,  sometimes, I managed to be displayed the same  carrier in the waterfall wrongly outside  of cursor , jus several khz below the cursor. 
 *WFM , shows a fixed received signal in the LCD cursor ,  but it keeps tuned ignoring any  freq. change . Examle , the same 90.2 FM broadcasting is tuned , even we move freq to 88 or 108 Mhz . .. It seems ignoring the tune freq. 

* In AM , still changing demodulation type ,  example  AM - USB - LSB ,... when coming back to the original mod type the good sound is lost, to be recover , we need to change freq. up and down .
---------------------------
So , it is a promising  very fast digital down convertion freq. , but we need more opinions , and  more stable platform .

strijar commented:
> WFM, shows a fixed received signal in the LCD cursor , but it keeps tuned ignoring any freq. change . Examle , the same 90.2 FM broadcasting is tuned , even we move freq to 88 or 108 Mhz . .. It seems ignoring the tune freq. ??? (I could replicate 100% of times)

I already fixed this - it just didn’t enter the binary image

> In AM , still changing demodulation type , example AM - USB - LSB

Fixed

gullradriel commented:
I will test it tonight
